File: src/main/java/gargoyle/calendar/util/reflect/Reflections.java
package gargoyle.calendar.util.reflect;

import java.lang.reflect.Constructor;
import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;
import java.text.MessageFormat;

public final class Reflections {

    private Reflections() {
    }


    public static <B> B newInstance(Class<B> configClass) {
        try {
            Constructor<B> constructor = configClass.getConstructor();
            return constructor.newInstance();
        } catch (InstantiationException | IllegalAccessException | InvocationTargetException | NoSuchMethodException e) {
            String configClassName = configClass.getName();
            throw new ReflectionException(MessageFormat.format("cannot instantiate class {0}", configClassName), e);
        }
    }

    public static Method getMethod(Class<?> aClass, String methodName) {
        try {
            return aClass.getMethod(methodName);
        } catch (NoSuchMethodException e) {
            throw new ReflectionException(e.getLocalizedMessage());
        }
    }
}

File: cal-core/src/main/kotlin/gargoyle/calendar/util/Reflections.kt
package gargoyle.calendar.util

import kotlin.reflect.KClass
import kotlin.reflect.KFunction
import kotlin.reflect.full.memberFunctions

object Reflections {
    fun <B : Any> newInstance(configClass: KClass<B>): B =
        configClass.constructors.first { it.parameters.isEmpty() }.call()

    fun getMethod(type: KClass<*>, methodName: String): KFunction<*> =
        type.memberFunctions.first { it.name == methodName }
}
